Wip/ported my blog to statiq (#4)
mholo65/mholo65: input/posts/2020-04-20-i-ported-my-blog-to-statiq.md

Title: I ported my blog to Statiq
Published: 2020-04-20
Tags: 
- .NET
- Statiq
- Static site generator
---
### Background

> About six months ago, I got assigned a task to write a convention-driven documentation generator at work. Our first thought was to use [Wyam](https://wyam.io/) for this, but after discussions with [Dave Glick](https://github.com/daveaglick/), he lured us into trying out this new shiny stuff he was working on, called [Statiq](https://statiq.dev/). Creating new stuff with Wyam at that point would have resulted in a rewrite once Statiq would be released, making Wyam obsolete.

Reviewer: ```suggestion
About six months ago, I got assigned a task to write a convention-driven documentation generator at work. Our first thought was to use [Wyam](https://wyam.io/) for this, but after discussions with [Dave Glick](https://github.com/daveaglick/), he lured us into trying a new shiny project he was working on called [Statiq](https://statiq.dev/). Creating a new site with Wyam at that point would've resulted in a rewrite once Statiq was released since it would make Wyam obsolete.
```

>
> -- <cite>[Statiq/Dave Glick](https://statiq.dev/)</cite>

Seems compelling, right? Additionally, since Statiq runs on .NET, you will have all the power of .NET in your static generator pipelines. This is what really makes it shine!

#### How Statiq works
_(The following section is copied from the official Statiq documentation and reflects the state of Statiq by the time this blog post was written)_
![Statiq Flow](/assets/images/statiq-flow.png){width=100%}

Statiq is powerful because it combines a few simple building blocks that can be rearranged and used in limitless combinations. Think of it like LEGO® for static generation.
* Content and data can come from a variety of sources including input files, databases, and services.
* Documents are created which each contain content and metadata.
* The documents are processed by pipelines.
* Each pipeline consists of one or more modules that manipulate the documents given to it by transforming, aggregating, filtering, or producing entirely new documents.
* The final output of each pipeline is made available to other pipeline and may be written to output files or deployed to hosting services.

### Motivation

> Back in December, when the documentation generator I was working on was feature complete, I wanted to continue using Statiq as it was still evolving. This was when I decided that my next project will be porting this blog to use Statiq. The old Wyam template that I was using worked, but it was something I had copy-pasted from [Gary Park's blog](https://github.com/gep13/gep13) and I had almost zero knowledge on what happens behind the scene. As most software developers, I do not trust magic, I need to know how stuff works. Therefore, my next project would be porting this blog to Statiq.

Reviewer: ```suggestion
Back in December, when the documentation generator I was working on became feature complete, I wanted to continue using Statiq as it continued to evolve. This was when I decided my next project would be porting this blog to use Statiq. The old Wyam template I was using worked, but it was something I copy-pasted from [Gary Park's blog](https://github.com/gep13/gep13) and I had almost zero knowledge on what happens behind the scenes. As with most software developers, I don't trust magic and need to know how stuff works. Therefore, it was decided my next project would be porting this blog to Statiq.
```

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">TODO: Rewrite my blog with <a href="https://twitter.com/statiqdev?ref_src=twsrc%5Etfw">@statiqdev</a> and then blog about it. I&#39;ve had the privilege to work with the framework for a work thingy and I see great potential in it. Wyam was pure magic to me, but Statiq I can understand 😀</p>&mdash; Martin Björkström (@mholo65) <a href="https://twitter.com/mholo65/status/1202304328465301512?ref_src=twsrc%5Etfw">December 4, 2019</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 

<br/>

> ### Attempt 1 - Porting Wyam Blog recipe to Statiq

Reviewer: ```suggestion
### Attempt 1 - Porting the Wyam Blog recipe to Statiq
```

> My first though was to port the old Wyam blog recipe and theme to Statiq. So, I cloned Wyam sources and started porting these to Statiq.

Reviewer: ```suggestion
My first thought was to port the old Wyam blog recipe and theme to Statiq, so I cloned the Wyam sources and started porting these to Statiq.
```

* [Wyam Blog Recipe](https://github.com/Wyamio/Wyam/tree/94a3f1ba258b7d1aaf4f9e55b222697698346396/src/recipes/Wyam.Blog)
* [Phantom Blog Theme](https://github.com/Wyamio/Wyam/tree/94a3f1ba258b7d1aaf4f9e55b222697698346396/themes/Blog/Phantom)

After a couple of hours working on this, which can be found [here](https://github.com/mholo65/mholo65/commits/feature/statiq), I quickly realized that this is very difficult and time consuming. And worst of all, I was porting code which I had no idea of what it was doing. The Wyam blog recipe was very extensible and I did not need many of the extension points it was offering. Some of the code would also have been better off just re-written in Statiq. Once again, I consulted Dave...

### Attempt 2 - Using Statiq.Web

> Once Dave was happy with [Statiq Framework](https://github.com/statiqdev/Statiq.Framework), which is the core of Statiq, he started working on the replacements for Wyam recipes, namely [Statiq.Web](https://github.com/statiqdev/Statiq.Web) and [Statiq.Docs](https://github.com/statiqdev/Statiq.Docs). So at the beginning of March I decided to give Statiq.Web a go (source code [here](https://github.com/mholo65/mholo65/commits/statiq.web)). I quickly faced same problem here as with [Attempt 1](#attempt-1-porting-wyam-blog-recipe-to-statiq). While the recipe was mostly ported, it still lacked themes, which led me into porting the old Wyam themes to Statiq. The result would have been much like using Wyam, i.e. magic which I did not understand :) So I wanted full control...

Reviewer: ```suggestion
Once Dave was happy with [Statiq Framework](https://github.com/statiqdev/Statiq.Framework), which is the core of Statiq, he started working on a replacement for the Wyam recipes, namely [Statiq Web](https://github.com/statiqdev/Statiq.Web) and [Statiq Docs](https://github.com/statiqdev/Statiq.Docs). So at the beginning of March I decided to give Statiq Web a go (source code [here](https://github.com/mholo65/mholo65/commits/statiq.web)). I quickly faced the same problems with that as with [Attempt 1](#attempt-1-porting-wyam-blog-recipe-to-statiq). While the recipe was partially ported, it still lacked themes, which led me into porting the old Wyam themes to Statiq. The result would have been much like using Wyam, i.e. magic which I did not understand :) So I wanted full control...
```

### Attempt 3 - Using Statiq.Framework

> As I had used [Statiq Framework](https://github.com/statiqdev/Statiq.Framework) previously for work related stuff, I was quite comfortable with this approach. I started by [contributing a Handlebars module to Statiq](https://github.com/statiqdev/Statiq.Framework/pull/90), because I prefer using Handlebars over Razor for simple templates. Once the Handlebars module was pulled in, I started porting my blog. The first thing I did was to take the HTML output from the old Wyam-powered blog and create handlebars templates from it. The next step was to create some pipelines. The basics of the pipelines are described in the following sections. To better understand the concept of Statiq pipelines and phases, please read [Statiq's offical documentation](https://statiq.dev/framework/concepts/pipelines). The full source for the pipelines that create my blog can be found on [GitHub](https://github.com/mholo65/mholo65/tree/master/src/Pipelines).

Reviewer: Remind me to come back in and double-check the Statiq docs links like the one to "pipelines" here. I'm finding as I fill them out that the current topic structure isn't quite right and I'll probably shuffle them around a bit later this week.

Author: Yep, thanks!

#### Blog post pipeline

> The main pipeline in my blog engine is the pipeline that handles the blog posts. It reads all blog posts, using the `./posts/*.md` glob pattern in the `input` phase. The `process` phase extracts metadata from Front Matter (the small YAML section on top of the document), renders markdown to html, generates excerpt and lastly changes the destination extension to `.html`. We will skip the `post process` phase for now and go straight to the `output` phase which basically just writes the processed documents to disk.

Reviewer: ```suggestion
The main pipeline in my blog engine is the pipeline that handles the blog posts. It reads all blog posts, using the `./posts/*.md` glob pattern in the `input` phase. The `process` phase extracts metadata from Front Matter (the small YAML section on top of the document), renders markdown to HTML, generates an excerpt, and lastly changes the destination extension to `.html`. We'll skip the `post process` phase for now and go straight to the `output` phase which basically just writes the processed documents to disk.
```

```csharp
public BlogPostPipeline()
{
    InputModules = new ModuleList
    {
        new ReadFiles("./posts/*.md")
    };

    ProcessModules = new ModuleList
    {
            new ExtractFrontMatter(new ParseYaml()),
            new RenderMarkdown()
                .UseExtensions(),
            new GenerateExcerpt(),
            new SetDestination(".html")
    };

    ...

    OutputModules = new ModuleList
    {
        new WriteFiles()
    };
}
```

#### Tags pipeline

> Every blog post contains a list of tags, defined in the Front Matter. These tags are displayed on the various pages, including the post itself, the index page (which shows the top-ten tags), and the tags page which list all tags. For each tag, the site also have a page which lists all blog posts tagged with the specific tag (see e.g. [`.NET`](/tags/net)). The tags pipeline creates the tag documents which is used on the various pages and which results in the separate tag pages. In order to get the tags from the blog posts, we specify that this pipeline has a dependency on the blog post pipeline. This will instruct Statiq to execute `process` phase of the blog post pipeline before executing the `process` phase of the tags pipeline and make the blog posts available inside the `process` phase.

Reviewer: ```suggestion
Every blog post contains a list of tags, defined in the Front Matter. These tags are displayed on the various pages, including the post itself, the index page (which shows the top-ten tags), and the tags page which list all tags. For each tag, the site also has a page which lists all blog posts tagged with the specific tag (see e.g. [`.NET`](/tags/net)). The tags pipeline creates the tag documents which is used on the various pages and which results in the separate tag pages. In order to get the tags from the blog posts, we specify that this pipeline has a dependency on the blog post pipeline. This will instruct Statiq to execute the `process` phase of the blog post pipeline before executing the `process` phase of the tags pipeline and make the blog posts available inside the `process` phase.
```

```csharp
public BlogPostPipeline()
{
    ...

    PostProcessModules = PostProcessModules.Prepend(
        new SetMetadata("template",
            Config.FromContext(async ctx =>
                await ctx.FileSystem.GetInputFile("_post.hbs").ReadAllTextAsync())),
        new RenderHandlebars("template")
            .WithModel(Config.FromDocument(async (doc, context) => new
            {
                title = doc.GetString(Keys.Title),
                date = doc.GetDateTime(FeedKeys.Published).ToLongDateString(),
                body = await doc.GetContentStringAsync(),
                tags = doc.GetList<string>("tags")
                    .OrderBy(x => x)
                    .Select(x => context.Outputs
                        .FromPipeline(nameof(TagsPipeline))
                        .First(tag => tag.GetString(Keys.GroupKey) == x))
                    .Select(x => x.AsTag(context))
            })),
        new SetContent(Config.FromDocument(x => x.GetString("template"))));

    ...
}
```

### Summary

> If you have made it this far, congratulations, now you hopefully know a little bit more about how Statiq works and the concept of pipelines, modules, and documents. For me, the biggest _aha moment_ when starting out with Statiq came when I realized that documents are just placeholders for metadata. Documents may or may not contain content, but they will always contain some metadata. Documents may or may not exist in the input directory, and the may or may not be written to the output directory, but they will always contain metadata.

Reviewer: ```suggestion
If you've made it this far, congratulations! Now you hopefully know a little bit more about how Statiq works and the concept of pipelines, modules, and documents. For me, the biggest _aha moment_ when starting out with Statiq came when I realized that documents are just placeholders for metadata. Documents may or may not contain content, but they will always contain some metadata. Documents may or may not exist in the input directory, and the may or may not be written to the output directory, but they will always contain metadata.
```

> Please note, that if your only intention is to create a blog, or port your existing Wyam-based blog to Statiq, this approach may not be what you wanted. You should therefore wait for [`Statiq.Web`](https://github.com/statiqdev/Statiq.Web/) to be released, and you'll get the same (or even better) experience that you can get with Wyam today. But if you, like me, likes tinkering then I definitely recommend creating you own site just using the modules found in [`Statiq.Framework`](https://github.com/statiqdev/Statiq.Framework). It is also very easy to create custom modules yourself, even though this does not contain any information about that. But stay tuned, who knows, maybe I will write a post about custom modules or shortcodes in the future.

Reviewer: ```suggestion
Please note, that if your only intention is to create a blog, or port your existing Wyam-based blog to Statiq, this approach may not be what you want. You should consider waiting for [Statiq Web](https://github.com/statiqdev/Statiq.Web/) to be released, and you'll get the same (or even better) experience that you can get with Wyam today. But if you're like me and like tinkering then I definitely recommend creating your own site just using the modules found in [Statiq Framework](https://github.com/statiqdev/Statiq.Framework). It's also very easy to create custom modules yourself, even though this post does not contain any information about that. But stay tuned, who knows, maybe I will write a post about custom modules or shortcodes in the future.
```
